nuxeo-opencmis-impl/src/main/java/org/nuxeo/ecm/core/opencmis/impl/client/NuxeoSession.java

```diff
@@ -254,6 +256,115 @@ public ObjectId createItem(Map<String, ?> properties, ObjectId folderId) {
         throw new CmisNotSupportedException();
     }
 
+    @Override
+    public ObjectId createPath(String newPath, Map<String, ?> properties) {
+        return createPath(null, newPath, properties);
+    }
+
+    @Override
+    public ObjectId createPath(ObjectId startFolderId, String newPath, Map<String, ?> properties) {
+        return createPath(startFolderId, newPath, properties, null, null, null);
+    }
+
+    @Override
+    public ObjectId createPath(String newPath, String typeId) {
+        return createPath(null, newPath, typeId);
+    }
+
+    @Override
+    public ObjectId createPath(ObjectId startFolderId, String newPath, String typeId) {
+        Map<String, Object> properties = Collections.singletonMap(PropertyIds.OBJECT_TYPE_ID, typeId);
+        return createPath(startFolderId, newPath, properties, null, null, null);
+    }
+
+    @Override
+    public ObjectId createPath(ObjectId startFolderId, String newPath, Map<String, ?> properties, List<Policy> policies,
+            List<Ace> addAces, List<Ace> removeAces) {
+        checkPath(newPath);
+        if (newPath.length() == 1) {
+            throw new CmisInvalidArgumentException("Cannot create root folder");
+        }
+        if (newPath.endsWith("/")) {
+            throw new CmisInvalidArgumentException("Path cannot end with a slash");
+        }
+        if (properties == null || properties.isEmpty()) {
+            throw new CmisInvalidArgumentException("Properties must not be empty");
+        }
+        if (!(properties.get(PropertyIds.OBJECT_TYPE_ID) instanceof String)) {
+            throw new CmisInvalidArgumentException("Property cmis:objectTypeId not set or invalid");
+        }
+        StringBuilder nextPath = new StringBuilder();
+        String[] segments;
+        ObjectId lastFolderId = null;
+        boolean create = false;
+
+        // check start folder
+        if (startFolderId != null && startFolderId.getId() != null) {
+            if (startFolderId instanceof Folder) {
+                Folder startFolder = (Folder) startFolderId;
+                if (!startFolder.isRootFolder()) {
+                    nextPath.append(startFolder.getPath());
+                    lastFolderId = startFolder;
+                }
+            } else {
+                ObjectData startFolderData = service.getObject(repositoryId, startFolderId.getId(),
+                        "cmis:objectId,cmis:baseTypeId,cmis:path", Boolean.FALSE, IncludeRelationships.NONE,
```

Comment:
Are there constants for these?

Reply:
There are yes, although this code was copied from Apache Chemistry's implementation and adapted. I'll use the constants.